code/scorpio/troll.dm

```diff
@@ -0,0 +1,120 @@
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+//------------------------------------------------------------------------------
+
+GLOBAL_LIST_EMPTY(gamer_words)
+
+/proc/add_gamer_words(config_value)
+	GLOB.gamer_words += splittext(config_value, ",")
+
+/proc/ban_and_kick_troll(context, message, word, mob_name, client/ban_me)
+	ban_troll(context, message, word, mob_name, ban_me)
+	kick_troll(ban_me)
+
+/proc/ban_troll(context, message, word, mob_name, client/ban_me)
+	// we can't ban if we don't have a database
+	if(!SSdbcore.IsConnected())
+		log_and_message_admins("Database connection failure when attempting to ban troll [ban_me].")
+		return FALSE
+
+	// build up all the things we need to record the ban
+	var/serverip = "[world.internet_address]:[world.port]"
+	var/bantype_str = "PERMABAN"
+	var/reason = "[ban_me] / ([mob_name]) used gamer word '[word]' in [context] message [message]"
+	var/job = ""
+	var/duration = -1 // PERMABAN
+	var/rounds = 0
+	var/ckey = "[ban_me.ckey]"
+	var/computerid = "[ban_me.computer_id]"
+	var/ip = "[ban_me.address]"
+	var/a_ckey = "ScorpioStationAI"
+	var/a_computerid = 0x7f000001
+	var/a_ip = "127.0.0.1"
+	var/who = english_list(GLOB.clients)
+	var/adminwho = english_list(GLOB.admins)
+
+	// add the ban to the database
+	var/datum/db_query/query_insert = SSdbcore.NewQuery({"
+		INSERT INTO [format_table_name("ban")] (`id`,`bantime`,`serverip`,`bantype`,`reason`,`job`,`duration`,`rounds`,`expiration_time`,`ckey`,`computerid`,`ip`,`a_ckey`,`a_computerid`,`a_ip`,`who`,`adminwho`,`edits`,`unbanned`,`unbanned_datetime`,`unbanned_ckey`,`unbanned_computerid`,`unbanned_ip`)
+		VALUES (null, Now(), :serverip, :bantype_str, :reason, :job, :duration, :rounds, Now() + INTERVAL :duration MINUTE, :ckey, :computerid, :ip, :a_ckey, :a_computerid, :a_ip, :who, :adminwho, '', null, null, null, null, null)
+	"}, list(
+		// Get ready for parameters
+		"serverip" = serverip,
+		"bantype_str" = bantype_str,
+		"reason" = reason,
+		"job" = job,
+		"duration" = (duration ? "[duration]" : "0"), // strings are important here
+		"rounds" = (rounds ? "[rounds]" : "0"), // and here
+		"ckey" = ckey,
+		"computerid" = computerid,
+		"ip" = ip,
+		"a_ckey" = a_ckey,
+		"a_computerid" = a_computerid,
+		"a_ip" = a_ip,
+		"who" = who,
+		"adminwho" = adminwho
+	))
+	if(!query_insert.warn_execute())
+		qdel(query_insert)
+		return FALSE
+	qdel(query_insert)
+	log_and_message_admins("ScorpioStationAI has added a [bantype_str] for [ckey] with the reason: \"[reason]\" to the ban database.")
```

Comment:
Should this be renamed to Troll Code? Just in case people get this confused with the AI player?

Reply:
Good call. Although normally only the admins would see it, the person who gets banned may confuse "ScorpioStationAI" with the person who was playing the AI on the station.

I've changed this artificial admin ckey to `Automated_Gamer_Word_Ban_System`. Hopefully nobody (admin or the banned) will get confused now.